// eas-cli/build/commands/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const path_1 = tslib_1.__importDefault(require("path"));
const runBuildAndSubmit_1 = require("../../build/runBuildAndSubmit");
const types_1 = require("../../build/types");
const EasCommand_1 = tslib_1.__importDefault(require("../../commandUtils/EasCommand"));
const log_1 = tslib_1.__importDefault(require("../../log"));
const platform_1 = require("../../platform");
const projectUtils_1 = require("../../project/projectUtils");
const json_1 = require("../../utils/json");
class Build extends EasCommand_1.default {
    async runAsync() {
        const { flags: rawFlags } = await this.parse(Build);
        if (rawFlags.json) {
            (0, json_1.enableJsonOutput)();
        }
        const flags = await this.sanitizeFlagsAsync(rawFlags);
        const projectDir = await (0, projectUtils_1.findProjectRootAsync)();
        await (0, runBuildAndSubmit_1.runBuildAndSubmitAsync)(projectDir, flags);
    }
    async sanitizeFlagsAsync(flags) {
        var _a, _b;
        const nonInteractive = flags['non-interactive'];
        if (!flags.local && flags.output) {
            core_1.Errors.error('--output is allowed only for local builds', { exit: 1 });
        }
        if (!flags.platform && nonInteractive) {
            core_1.Errors.error('--platform is required when building in non-interactive mode', { exit: 1 });
        }
        if (flags.json && !nonInteractive) {
            core_1.Errors.error('--json is allowed only when building in non-interactive mode', { exit: 1 });
        }
        const requestedPlatform = await (0, platform_1.selectRequestedPlatformAsync)(flags.platform);
        if (flags.local) {
            if (flags['auto-submit'] || flags['auto-submit-with-profile'] !== undefined) {
                // TODO: implement this
                core_1.Errors.error('Auto-submits are not yet supported when building locally', { exit: 1 });
            }
            if (requestedPlatform === platform_1.RequestedPlatform.All) {
                core_1.Errors.error('Builds for multiple platforms are not supported with flag --local', {
                    exit: 1,
                });
            }
            else if (process.platform !== 'darwin' && requestedPlatform === platform_1.RequestedPlatform.Ios) {
                core_1.Errors.error('Unsupported platform, macOS is required to build apps for iOS', { exit: 1 });
            }
        }
        if (flags['skip-credentials-check']) {
            log_1.default.warnDeprecatedFlag('skip-credentials-check', 'Build credentials validation is always skipped with the --non-interactive flag. You can also skip interactively.');
            log_1.default.newLine();
        }
        if (flags['skip-project-configuration']) {
            log_1.default.warnDeprecatedFlag('skip-project-configuration', 'Automatic configuration of native code is no longer optional.');
            log_1.default.newLine();
        }
        const message = flags['message'];
        if (message && message.length > 1024) {
            core_1.Errors.error('Message cannot be longer than 1024 characters.', { exit: 1 });
        }
        const profile = flags['profile'];
        return {
            requestedPlatform,
            profile,
            nonInteractive,
            localBuildOptions: flags['local']
                ? {
                    enable: true,
                    verbose: true,
                    artifactPath: flags.output && path_1.default.resolve(process.cwd(), flags.output),
                }
                : {
                    enable: false,
                },
            wait: flags['wait'],
            clearCache: flags['clear-cache'],
            json: flags['json'],
            autoSubmit: flags['auto-submit'] || flags['auto-submit-with-profile'] !== undefined,
            submitProfile: (_a = flags['auto-submit-with-profile']) !== null && _a !== void 0 ? _a : profile,
            userInputResourceClass: (_b = flags['resource-class']) !== null && _b !== void 0 ? _b : types_1.UserInputResourceClass.DEFAULT,
            message,
        };
    }
}
exports.default = Build;
Build.description = 'start a build';
Build.flags = {
    platform: core_1.Flags.enum({
        char: 'p',
        options: ['android', 'ios', 'all'],
    }),
    'skip-credentials-check': core_1.Flags.boolean({
        default: false,
        hidden: true,
    }),
    json: core_1.Flags.boolean({
        description: 'Enable JSON output, non-JSON messages will be printed to stderr',
        default: false,
    }),
    'skip-project-configuration': core_1.Flags.boolean({
        default: false,
        hidden: true,
    }),
    profile: core_1.Flags.string({
        description: 'Name of the build profile from eas.json. Defaults to "production" if defined in eas.json.',
        helpValue: 'PROFILE_NAME',
    }),
    'non-interactive': core_1.Flags.boolean({
        default: false,
        description: 'Run command in non-interactive mode',
    }),
    local: core_1.Flags.boolean({
        default: false,
        description: 'Run build locally [experimental]',
    }),
    output: core_1.Flags.string({
        description: 'Output path for local build',
    }),
    wait: core_1.Flags.boolean({
        default: true,
        allowNo: true,
        description: 'Wait for build(s) to complete',
    }),
    'clear-cache': core_1.Flags.boolean({
        default: false,
        description: 'Clear cache before the build',
    }),
    'auto-submit': core_1.Flags.boolean({
        default: false,
        description: 'Submit on build complete using the submit profile with the same name as the build profile',
        exclusive: ['auto-submit-with-profile'],
    }),
    'auto-submit-with-profile': core_1.Flags.string({
        description: 'Submit on build complete using the submit profile with provided name',
        helpValue: 'PROFILE_NAME',
        exclusive: ['auto-submit'],
    }),
    'resource-class': core_1.Flags.enum({
        options: Object.values(types_1.UserInputResourceClass),
        hidden: true,
        description: 'The instance type that will be used to run this build [experimental]',
    }),
    message: core_1.Flags.string({
        char: 'm',
        description: 'A short message describing the build',
    }),
};
